### ISSUE TITLE
Client: App max_concurrent causes continuous work fetch and too many tasks

### ISSUE THREAD

**Issue by sam8641** [BOINC/boinc#4322]
**Describe the bug**
Continuous work fetch going way past the configured number of days of work. This happens when using app max_concurrent, an amount less then the number of tasks received for this app, and the tasks for this app is the oldest and/or have longest run time. Probably an issue with rr_simulation.

**Steps To Reproduce**
1. Join/run one project website with multiple available apps. (World Community Grid)
2. BOINC settings, store 0 days, store up to 0.2 additional days.
3. Website settings, limit task to one app. (Africa Rainfull Project / ARP1). In World community grid, go to Settings, Device manager, profile, and also increase the project/app limit to 16 or higher or unlimited.
4. After 2 or more tasks are gotten for the same app, switch to another app using website settings.
5. Use app_config.xml, limit max_concurrent on arp1 to 1, or less then the number of arp1 tasks in the BOINC client.

**Expected behavior**
Limit tasks with no work fetch when enough of other apps are in BOINC client.

**System Information**
 - OS: Windows 10
 - BOINC Version: 7.16.11

**Additional context**
Log shows something for rr_simulation, with max_concurrent=12 on Ryzen 3900x 12 core 24 threads. The rr_simulation stops after only 12 ARP1 and fails to see 400 MCM1 tasks. This causes work fetch to keep fetching more new tasks every 2 minutes non-stop, way past the configured configured number of days of work. All 24 threads are fully used with 12 ARP1 plus 12 MCM1/MIP1.

Some log: [boinc_max_concurrent12.txt](https://github.com/BOINC/boinc/files/6295109/boinc_max_concurrent12.txt)

**Comment by cwallbaum** [comment]
See #4117

**Comment by MaxMaximus** [comment]
I see this bug often too. It happens if max_concurrent is used on app OR project level. Both use cases cause similar error of infinite work downloading loop ignoring work cache setting.
  
I run rr_sim debug  and confirm - it happens due to a wrong calculation of amount of work BOINC already have. 
rr_sim shows than BOINC stops calculating WUs for an app (OR project if max_concurrent is set on project level) after counting only up to max_concurrent WUs. Ignoring all others WUs after reaching max_concurrent.

It is a clear bug because max_concurrent is limiting only number of jobs to run in parallel (simultaneously). NOT a total number of jobs to run for an app/project.

For example. 
I use max_concurrent to limit Rosett@Home to run max 4 WUs in parallel (on project level in my case).
Each WU from R@H run for a ~8 hours and have a deadline of 3 days
I use 1 day work cache setting, so its should be about 12 WUs in work queue to match 1 day setting on 24/7 client.
But BOINC count only 4 WUs and "thinks" it have work only for ~8 hours in queue while setting is set to 1 day .
So it starts downloading WUs  in infinite loop (hundreds of WUs until it stopped by server side limit/quota of number of task in progress per one host), because due to bug it still count only 4 WUs in work cache regardless of how many it actually has. 

And its not a minor bug. Because it leads not just to an "excessive" work cache (exceeding users settings and expectation) but download of huge amounts of WUs most of which will miss dedlline and will be aborted. 

It puts stress on projects servers, "burn" client traffics on limited/metered internet connections, often leads to lost of significant amounts of computation (because of task reported too late and discarded - BOINC automatically abort task with missed deadline only if it was not started before deadline, but if it was already starded - keeps computing task with already missed deadline), puts client often into "panic" (high priority) mode due to missing deadlines and produce havoc on multi-project clients.

**Comment by mmstick** [comment]
Had this happen on two machines with the following `app_config.xml` on Rosetta@home:

```xml
<app_config>
  <app>
    <name>rosetta</name>
    <max_concurrent>12</max_concurrent>
  </app>
  <app>
    <name>rosetta_python_projects</name>
    <max_concurrent>4</max_concurrent>
  </app>
</app_config>
```

Went to eat for 15 minutes and both systems had fetched 500 tasks each and were continually fetching 12 new tasks every 30 seconds.

**Comment by davidpanderson** [comment]
Can you reproduce this in the client emulator?
https://boinc.berkeley.edu/sim_web.php

That would make it much easier for me to debug it.

**Comment by MaxMaximus** [comment]
There I reproduced one of the cases in sim:
https://boinc.berkeley.edu/scenarios/192/simulations/1/log.txt

Note it sim with actual configs  from my crunching machine running multiple projects. 
Easiest way to reproduce bug is to use just a single project with max_concurrent settings PLUS relative large work cache (exceeding amount of work  which can be run in one cycle - eg if max_concurrent allows 8 task, cache >8 tasks needed to trigger bug). Its enough to trigger it.
But i do not want to detach from all other projects and currently do not have a spare machine to test with new/clean BOINC installition.

In this more complex case this multiple projects, and only one of it have a max_concurrent limit in app_info (in my case it is a Rosetta@Home) all works fine while BOINC can fill the work queue (in this example set to 1.2 days of work) with tasks from other non limited by max_concurrent setting projects.
So i set other projects to NoNewTask before starting simulation. One day of sim all work OK, while BOINC work trough task from other projects. But is triggered at 1 day 00:51:30 point when BOINC is start running out of other task and tries to fill cache with work from R@H.
And failed - it get about 1000 R@H tasks in next 7 hours in non stop cycle. While each task takes 8 hours to compute, so with max concurrent limit = 4 it can finish about 3*24/8*4 = 36 tasks before deadline, all other (~960) will fail due to missed deadline.

**Comment by davidpanderson** [comment]
Thanks - I'll take a look at it.

**Comment by davidpanderson** [comment]
This is fixed by #4592; the simulator is running this version.
